engine/runtime/src/main/java/org/enso/interpreter/node/typecheck/LazyCheckRootNode.java

```diff
@@ -0,0 +1,46 @@
+package org.enso.interpreter.node.typecheck;
+
+import com.oracle.truffle.api.TruffleLanguage;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.nodes.RootNode;
+import org.enso.interpreter.node.BaseNode;
+import org.enso.interpreter.node.callable.thunk.ThunkExecutorNode;
+import org.enso.interpreter.runtime.callable.argument.ArgumentDefinition;
+import org.enso.interpreter.runtime.callable.function.Function;
+import org.enso.interpreter.runtime.callable.function.FunctionSchema;
+
+final class LazyCheckRootNode extends RootNode {
+
+  @Child private ThunkExecutorNode evalThunk;
+  @Child private TypeCheckValueNode check;
+  static final FunctionSchema SCHEMA =
+      FunctionSchema.newBuilder()
+          .argumentDefinitions(
+              new ArgumentDefinition(
+                  0, "delegate", null, null, ArgumentDefinition.ExecutionMode.EXECUTE))
+          .hasPreapplied(true)
+          .build();
+
+  LazyCheckRootNode(TruffleLanguage<?> language, TypeCheckValueNode check) {
+    super(language);
+    this.check = check;
+    this.evalThunk = ThunkExecutorNode.build();
+  }
+
+  Function wrapThunk(Function thunk) {
+    return new Function(getCallTarget(), thunk.getScope(), SCHEMA, new Object[] {thunk}, null);
+  }
+
+  @Override
+  public Object execute(VirtualFrame frame) {
+    org.enso.interpreter.runtime.state.State state =
```

Comment:
Using FQN in non-generated code when there are no ambiguous symbols seems really weird. Can't your IDE import these symbols automatically for you? Same comment for `java.lang.Object`.

Reply:
My IDE does not like `var` keyword during refactorings and replaces it by FQN. Annoying.

Reply:
Fixed in 4bc57eff4eb173c8b8a53a2e7d3f0600da9da189